File: frontend/lib/api/approvals.ts
// SPEC-005 승인 게이트 typed client — WORK-005 Phase 3.
// backend: app/schemas/approval.py 계약 기준. 후보 원장 state는 5개 enum,
// reanalysis_status는 표시용 파생값(원장 미저장 — DEC-022).

import { api, ApiError } from "./client";

// ── types ────────────────────────────────────────────────────────────────────

export type CandidateState =
  | "pending"
  | "stale"
  | "approved"
  | "rejected"
  | "blocked";

export type ReanalysisStatus =
  | "reanalyzing"
  | "new_candidate_ready"
  | "reanalysis_failed";

export type ReadCapability = "content_read" | "metadata_only";

export type RelationCandidateState =
  | "pending"
  | "unresolved"
  | "approved"
  | "removed";

export interface ApprovalCandidateSummary {
  id: number;
  document_id: number;
  drive_name: string;
  drive_state: string;
  state: CandidateState;
  reanalysis_status: ReanalysisStatus | null;
  read_capability: ReadCapability;
  stale_reason: string | null;
  blocked_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApprovalCandidateListResponse {
  candidates: ApprovalCandidateSummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface RelationCandidateItem {
  id: number;
  source_document_id: number;
  raw_label: string;
  suggested_relation_type:
    | "related"
    | "references"
    | "supersedes"
    | "duplicate_candidate";
  target_document_id: number | null;
  target_drive_name: string | null;
  state: RelationCandidateState;
  created_at: string;
  updated_at: string;
}

// AI candidate_metadata.resolution (backend services/classification._resolve_output)
export interface CandidateResolution {
  document_type_id: number | null;
  document_type_is_new: boolean;
  owning_department_node_id: number | null;
  created_department_node_id: number | null;
  organization_path_node_ids: (number | null)[];
  tree_path_node_ids: (number | null)[];
  read_department_node_ids: (number | null)[];
  related_department_node_ids: (number | null)[];
  unresolved_fields: string[];
  needs_admin_fix: boolean;
}

export interface CandidateMetadata {
  document_type?: string;
  created_department?: string | null;
  owning_department?: string;
  physical_tree_path?: { organization_path: string[]; tree_path: string[] };
  related_departments?: string[];
  related_products?: string[];
  summary?: string | null;
  sensitivity?: "normal" | "sensitive";
  policy_preset?: string | null;
  read_policy?: {
    read_roles: string[];
    read_departments: string[];
    read_positions: string[];
    access_logic: "ANY" | "ALL" | "PRESET";
  };
  reasons?: string[];
  resolution?: CandidateResolution;
  [key: string]: unknown;
}

export interface ApprovalCandidateDetail extends ApprovalCandidateSummary {
  drive_web_url: string | null;
  candidate_metadata: CandidateMetadata;
  candidate_fingerprint: Record<string, unknown>;
  current_fingerprint: Record<string, unknown>;
  fingerprint_match: boolean;
  approved_by: number | null;
  approved_at: string | null;
  relation_candidates: RelationCandidateItem[];
}

export interface ApprovalPayload {
  document_type_id: number;
  created_department_node_id: number | null;
  owning_department_node_id: number;
  physical_tree_path: { organization_path: number[]; tree_path: number[] };
  related_department_node_ids: number[];
  related_products: string[];
  summary: string | null;
  read_roles: string[];
  read_departments: number[];
  read_positions: string[];
  access_logic: "ANY" | "ALL" | "PRESET";
  sensitivity: "normal" | "sensitive";
  policy_preset: string | null;
}

export interface ApprovedMetadata {
  document_id: number;
  document_type_id: number | null;
  owning_department_node_id: number | null;
  organization_path: number[] | null;
  tree_path: number[] | null;
  access_logic: string;
  sensitivity: string;
  policy_preset: string | null;
  summary: string | null;
}

export interface ApprovalResultResponse {
  candidate: ApprovalCandidateSummary;
  document: ApprovedMetadata;
  idempotent: boolean;
}

export interface AdminDocumentTypeItem {
  id: number;
  name: string;
  normalized_name: string;
}

export interface RelationRematchResponse {
  candidate: RelationCandidateItem;
  suggested_target_document_id: number | null;
  suggested_target_drive_name: string | null;
}

// 민감 preset 후보 목록 (backend core/policy_presets — policy.md SoT, DEC-017)
export const POLICY_PRESETS = [
  "HR_RESTRICTED",
  "CONTRACT_RESTRICTED",
  "FINANCE_RESTRICTED",
  "SECURITY_RESTRICTED",
  "LEGAL_RESTRICTED",
] as const;

// ── SPEC-005 UX 문구 ─────────────────────────────────────────────────────────

export const CANDIDATE_STATE_LABEL: Record<CandidateState, string> = {
  pending: "pending",
  stale: "stale",
  approved: "approved",
  rejected: "rejected",
  blocked: "blocked",
};

export const REANALYSIS_STATUS_LABEL: Record<ReanalysisStatus, string> = {
  reanalyzing: "재분석 중",
  new_candidate_ready: "새 후보 준비됨",
  reanalysis_failed: "재분석 실패",
};

export const STALE_MESSAGE =
  "Drive 파일이 변경되어 이 후보는 승인할 수 없습니다.";
export const METADATA_ONLY_MESSAGE =
  "본문 분석 없이 Drive 정보만으로 생성된 후보입니다.";
export const BLOCKED_MESSAGE = "현재 문서 상태에서는 승인할 수 없습니다.";
export const EMPTY_QUEUE_MESSAGE = "승인할 후보가 없습니다.";

// ── SPEC-005 Case Matrix 프론트 출력 ─────────────────────────────────────────

const SPEC005_MESSAGES: Record<string, string> = {
  FORBIDDEN_ADMIN_ONLY: "관리자만 사용할 수 있습니다.",
  CANDIDATE_NOT_FOUND: "후보를 찾을 수 없습니다.",
  CANDIDATE_NOT_PENDING: "승인할 수 없는 후보 상태입니다.",
  CANDIDATE_STALE: "Drive 파일이 변경되어 다시 분석해야 합니다.",
  DOCUMENT_UNAVAILABLE: "현재 문서 상태에서는 승인할 수 없습니다.",
  DOCUMENT_TYPE_DUPLICATE: "이미 존재하는 문서종류입니다.",
  DOCUMENT_TYPE_NOT_FOUND: "문서종류를 카탈로그에서 찾을 수 없습니다.",
  INVALID_TREE_PATH: "문서 위치를 확인하세요.",
  INVALID_ACCESS_POLICY: "접근 권한 설정을 확인하세요.",
  RELATION_TARGET_REQUIRED: "대상 문서를 선택하거나 보류하세요.",
  REANALYSIS_FAILED: "재분석 요청에 실패했습니다.",
  ORG_NODE_NOT_FOUND: "조직을 찾을 수 없습니다.",
  ORG_NODE_INACTIVE: "비활성 조직은 새 귀속 대상으로 선택할 수 없습니다.",
  DOCUMENT_NOT_FOUND: "문서를 찾을 수 없습니다.",
};

/** ApiError → 한국어 카피 (SPEC-005 Case Matrix '프론트 출력'). */
export function spec005Message(err: unknown, fallback: string): string {
  if (
    err instanceof ApiError &&
    err.errorCode &&
    SPEC005_MESSAGES[err.errorCode]
  ) {
    return SPEC005_MESSAGES[err.errorCode];
  }
  return fallback;
}

// ── calls ────────────────────────────────────────────────────────────────────

export function fetchApprovalCandidates(params?: {
  state?: CandidateState;
  read_capability?: ReadCapability;
  limit?: number;
  offset?: number;
}): Promise<ApprovalCandidateListResponse> {
  const query = new URLSearchParams();
  if (params?.state) query.set("state", params.state);
  if (params?.read_capability)
    query.set("read_capability", params.read_capability);
  query.set("limit", String(params?.limit ?? 100));
  query.set("offset", String(params?.offset ?? 0));
  return api<ApprovalCandidateListResponse>(
    `/admin/approval-candidates?${query.toString()}`,
  );
}

export function fetchApprovalCandidate(
  candidateId: number,
): Promise<ApprovalCandidateDetail> {
  return api<ApprovalCandidateDetail>(
    `/admin/approval-candidates/${candidateId}`,
  );
}

export function approveCandidate(
  candidateId: number,
  payload: ApprovalPayload,
): Promise<ApprovalResultResponse> {
  return api<ApprovalResultResponse>(
    `/admin/approval-candidates/${candidateId}/approve`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
}

export function rejectCandidate(
  candidateId: number,
  reason?: string,
): Promise<ApprovalCandidateSummary> {
  return api<ApprovalCandidateSummary>(
    `/admin/approval-candidates/${candidateId}/reject`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ reason: reason ?? null }),
    },
  );
}

export function reanalyzeCandidate(candidateId: number): Promise<unknown> {
  return api<unknown>(`/admin/approval-candidates/${candidateId}/reanalyze`, {
    method: "POST",
  });
}

export function fetchAdminDocumentTypes(): Promise<AdminDocumentTypeItem[]> {
  return api<{ document_types: AdminDocumentTypeItem[] }>(
    "/admin/document-types",
  ).then((res) => res.document_types);
}

export function createAdminDocumentType(
  name: string,
): Promise<AdminDocumentTypeItem> {
  return api<AdminDocumentTypeItem>("/admin/document-types", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
}

export function resolveRelationCandidate(
  candidateId: number,
  targetDocumentId: number,
): Promise<RelationCandidateItem> {
  return api<RelationCandidateItem>(
    `/admin/relation-candidates/${candidateId}/resolve`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ target_document_id: targetDocumentId }),
    },
  );
}

export function holdRelationCandidate(
  candidateId: number,
): Promise<RelationCandidateItem> {
  return api<RelationCandidateItem>(
    `/admin/relation-candidates/${candidateId}/hold`,
    { method: "POST" },
  );
}

export function removeRelationCandidate(
  candidateId: number,
): Promise<RelationCandidateItem> {
  return api<RelationCandidateItem>(
    `/admin/relation-candidates/${candidateId}/remove`,
    { method: "POST" },
  );
}

export function rematchRelationCandidate(
  candidateId: number,
): Promise<RelationRematchResponse> {
  return api<RelationRematchResponse>(
    `/admin/relation-candidates/${candidateId}/rematch`,
    { method: "POST" },
  );
}

File: backend/app/api/routers/approvals.py
"""승인 게이트 admin 라우트 — WORK-005 (SPEC-005 API Contract).

- GET  /admin/approval-candidates          : 후보 큐 (state/read_capability 필터)
- GET  /admin/approval-candidates/{id}     : 후보 상세 (current fingerprint 동봉)
- POST /admin/approval-candidates/{id}/approve   : 승인 (재검사 → 반영, 멱등)
- POST /admin/approval-candidates/{id}/reject    : 거절 (pending/stale)
- POST /admin/approval-candidates/{id}/reanalyze : 수동 재분석 (WORK-004 위임)

에러봉투 {detail:{error_code,message}} — SPEC-005 Case Matrix.
reanalysis_status는 표시용 파생값으로 응답에만 실린다 (DEC-022).
"""

from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, require_admin_only
from app.api.errors import SPEC005_ERRORS, spec005_http_error
from app.api.routers.classification import _job_out
from app.dtos.user import UserDTO
from app.models.candidate import MetadataCandidate, RelationCandidate
from app.models.document import Document
from app.schemas.approval import (
    ApprovalCandidateDetail,
    ApprovalCandidateListResponse,
    ApprovalCandidateSummary,
    ApprovalPayload,
    ApprovalResultResponse,
    ApprovedMetadataOut,
    RejectPayload,
    RelationCandidateOut,
)
from app.schemas.classification import ClassificationJobOut
from app.services.ai_jobs import fingerprint_key
from app.services.approval import ApprovalService

router = APIRouter(tags=["approvals"])


def _summary_out(
    candidate: MetadataCandidate,
    document: Document,
    reanalysis_status: str | None,
) -> ApprovalCandidateSummary:
    return ApprovalCandidateSummary(
        id=candidate.id,
        document_id=candidate.document_id,
        drive_name=document.drive_name,
        drive_state=document.drive_state,
        state=candidate.state,  # type: ignore[arg-type]
        reanalysis_status=reanalysis_status,  # type: ignore[arg-type]
        read_capability=candidate.read_capability,  # type: ignore[arg-type]
        stale_reason=candidate.reason if candidate.state == "stale" else None,
        blocked_reason=candidate.reason if candidate.state == "blocked" else None,
        created_at=candidate.created_at,
        updated_at=candidate.updated_at,
    )


def _relation_out(
    rel: RelationCandidate, target_drive_name: str | None
) -> RelationCandidateOut:
    return RelationCandidateOut(
        id=rel.id,
        source_document_id=rel.source_document_id,
        raw_label=rel.raw_label,
        suggested_relation_type=rel.suggested_relation_type,  # type: ignore[arg-type]
        target_document_id=rel.target_document_id,
        target_drive_name=target_drive_name,
        state=rel.state,  # type: ignore[arg-type]
        created_at=rel.created_at,
        updated_at=rel.updated_at,
    )


def _approved_out(document: Document, related_ids: list[int]) -> ApprovedMetadataOut:
    return ApprovedMetadataOut(
        document_id=document.id,
        document_type_id=document.document_type_id,
        created_department_node_id=document.created_department_node_id,
        owning_department_node_id=document.owning_department_node_id,
        organization_path=document.organization_path,
        tree_path=document.tree_path,
        related_department_node_ids=related_ids,
        related_products=document.related_products,
        read_roles=document.read_roles,
        read_departments=document.read_departments,
        read_positions=document.read_positions,
        access_logic=document.access_logic,
        sensitivity=document.sensitivity,
        policy_preset=document.policy_preset,
        summary=document.summary,
    )


@router.get(
    "/admin/approval-candidates", response_model=ApprovalCandidateListResponse
)
async def list_approval_candidates(
    state: Literal["pending", "stale", "approved", "rejected", "blocked"] | None = Query(
        default=None
    ),
    read_capability: Literal["content_read", "metadata_only"] | None = Query(
        default=None
    ),
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    _admin: UserDTO = Depends(require_admin_only),
    session: AsyncSession = Depends(get_db),
) -> ApprovalCandidateListResponse:
    rows, total = await ApprovalService(session).list_candidates(
        state=state, read_capability=read_capability, limit=limit, offset=offset
    )
    return ApprovalCandidateListResponse(
        candidates=[
            _summary_out(candidate, document, status)
            for candidate, document, status in rows
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get(
    "/admin/approval-candidates/{candidate_id}",
    response_model=ApprovalCandidateDetail,
)
async def get_approval_candidate(
    candidate_id: int,
    _admin: UserDTO = Depends(require_admin_only),
    session: AsyncSession = Depends(get_db),
) -> ApprovalCandidateDetail:
    try:
        candidate, document, status, relations = await ApprovalService(
            session
        ).get_candidate(candidate_id)
    except SPEC005_ERRORS as exc:
        raise spec005_http_error(exc)
    summary = _summary_out(candidate, document, status)
    return ApprovalCandidateDetail(
        **summary.model_dump(),
        drive_web_url=document.drive_web_url,
        candidate_metadata=candidate.candidate_metadata,
        candidate_fingerprint=candidate.candidate_fingerprint,
        current_fingerprint=document.drive_fingerprint,
        fingerprint_match=fingerprint_key(candidate.candidate_fingerprint)
        == fingerprint_key(document.drive_fingerprint),
        approved_by=candidate.approved_by,
        approved_at=candidate.approved_at,
        relation_candidates=[
            _relation_out(rel, target_name) for rel, target_name in relations
        ],
    )


@router.post(
    "/admin/approval-candidates/{candidate_id}/approve",
    response_model=ApprovalResultResponse,
)
async def approve_candidate(
    candidate_id: int,
    payload: ApprovalPayload,
    admin: UserDTO = Depends(require_admin_only),
    session: AsyncSession = Depends(get_db),
) -> ApprovalResultResponse:
    service = ApprovalService(session)
    try:
        candidate, document, related_ids, idempotent = await service.approve(
            candidate_id, payload, admin=admin
        )
    except SPEC005_ERRORS as exc:
        raise spec005_http_error(exc)
    return ApprovalResultResponse(
        candidate=_summary_out(candidate, document, None),
        document=_approved_out(document, related_ids),
        idempotent=idempotent,
    )


@router.post(
    "/admin/approval-candidates/{candidate_id}/reject",
    response_model=ApprovalCandidateSummary,
)
async def reject_candidate(
    candidate_id: int,
    payload: RejectPayload | None = None,
    admin: UserDTO = Depends(require_admin_only),
    session: AsyncSession = Depends(get_db),
) -> ApprovalCandidateSummary:
    service = ApprovalService(session)
    try:
        candidate = await service.reject(
            candidate_id,
            admin=admin,
            reason=payload.reason if payload else None,
        )
        _, document, status, _ = await service.get_candidate(candidate.id)
    except SPEC005_ERRORS as exc:
        raise spec005_http_error(exc)
    return _summary_out(candidate, document, status)


@router.post(
    "/admin/approval-candidates/{candidate_id}/reanalyze",
    response_model=ClassificationJobOut,
    status_code=202,
)
async def reanalyze_candidate(
    candidate_id: int,
    admin: UserDTO = Depends(require_admin_only),
    session: AsyncSession = Depends(get_db),
) -> ClassificationJobOut:
    try:
        job = await ApprovalService(session).reanalyze(candidate_id, admin=admin)
    except SPEC005_ERRORS as exc:
        raise spec005_http_error(exc)
    return _job_out(job)
